Crash on WebCore::FrameLoader::effectiveReferrerPolicy() after 274396@main on ARM64 with GCC -O3

clopez opened the pull request:
#### bc889156b6fb70f9bde783b237fa4abb10e450ed
<pre>
Crash on WebCore::FrameLoader::effectiveReferrerPolicy() after 274396@main on ARM64 with GCC -O3
<a href="https://bugs.webkit.org/show_bug.cgi?id=273703">https://bugs.webkit.org/show_bug.cgi?id=273703</a>

Reviewed by Michael Catanzaro.

This workarounds a bug on GCC (affecting at least GCC 12, 13, 14, 15-nightly)
where the compiler produces wrong code on ARM64 at certain optimization levels
(like -O3) which seems related to lambdas and variants.

This wrong code later leads to a crash when the browser starts at
WebCore::FrameLoader::effectiveReferrerPolicy()

* Source/WebCore/page/Page.cpp:
(WebCore::createMainFrame):

Canonical link: <a href="https://commits.webkit.org/279066@main">https://commits.webkit.org/279066@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5d2bf7ae7754b351a817ca3bc2ba6c0a88ffb22

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52384 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55658 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3107 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54689 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2806 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42594 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1981 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54480 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29354 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45214 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23675 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26571 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2489 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1266 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48463 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2639 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/interactive-content.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57254 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2668 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49986 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28743 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45334 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49236 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29655 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7679 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28488 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (changes requested):
If this approach is to be taken, Page.cpp would need to be added to the WebCore target in Xcode.
I'd like to know more before approving this.  Last time I saw something like this it was undefined behavior from uninitialized memory.  If it's not, we should reference a gcc bug.

cdumez commented on Source/WebCore/page/Page.cpp:
```diff
@@ -290,7 +290,7 @@ static constexpr OptionSet<ActivityState> pageInitialActivityState()
 }
 
-static Ref<Frame> createMainFrame(Page& page, PageConfiguration::ClientCreatorForMainFrame&& clientCreator, RefPtr<Frame> mainFrameOpener, FrameIdentifier identifier)
+static Ref<Frame> createMainFrame(Page& page, PageConfiguration::ClientCreatorForMainFrame&& clientCreator, RefPtr<Frame>&& mainFrameOpener, FrameIdentifier&& identifier)
```
FrameIdentifier is OK to pass by value, this is a small type.

mcatanzaro replied:
This is a workaround for what we suspect is a GCC bug, though. It crashes if passed by value.

mcatanzaro replied:
Maybe add a FIXME comment with a link to this bug report. Otherwise, somebody is likely going to change it right back to pass by value.

clopez replied:
The crash is still there if I pass by value FrameIdentifier, only passing by reference both parameters fixes the issue

cdumez replied:
But that's not a "FIX", we should not do this. Either it's a compiler bug or we're working around a WebKit bug, which we should identify.

clopez replied:
Ok. Fair enough. It took me quite a while but finally I managed to create a simplified test case and I have reported the bug to GCC here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115135

After seeing that I'm almost sure this is a bug in GCC and not something wrong with the WebKit code.

I can add a comment on the code pointing to the GCC bug.

cdumez commented on Source/WebCore/page/Page.cpp:
```diff
@@ -327,7 +327,7 @@ Page::Page(PageConfiguration&& pageConfiguration)
     , m_backForwardController(makeUniqueRef<BackForwardController>(*this, WTFMove(pageConfiguration.backForwardClient)))
     , m_editorClient(WTFMove(pageConfiguration.editorClient))
-    , m_mainFrame(createMainFrame(*this, WTFMove(pageConfiguration.clientCreatorForMainFrame), WTFMove(pageConfiguration.mainFrameOpener), pageConfiguration.mainFrameIdentifier))
+    , m_mainFrame(createMainFrame(*this, WTFMove(pageConfiguration.clientCreatorForMainFrame), WTFMove(pageConfiguration.mainFrameOpener), WTFMove(pageConfiguration.mainFrameIdentifier)))
```
Second WTFMove() is not needed.

mcatanzaro commented on Source/WebCore/page/Page.cpp:
```diff
@@ -291,7 +291,14 @@ static constexpr OptionSet<ActivityState> pageInitialActivityState()
 }
 
-static Ref<Frame> createMainFrame(Page& page, PageConfiguration::ClientCreatorForMainFrame&& clientCreator, RefPtr<Frame> mainFrameOpener, FrameIdentifier identifier)
+// workaround for GCC bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115135
```
```suggestion
// FIXME: workaround for GCC bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115135
```

clopez committed:
Crash on WebCore::FrameLoader::effectiveReferrerPolicy() after 274396@main on ARM64 with GCC -O3
https://bugs.webkit.org/show_bug.cgi?id=273703

Reviewed by Michael Catanzaro.

This workarounds a bug on GCC (affecting at least GCC 12, 13, 14, 15-nightly)
where the compiler produces wrong code on ARM64 at certain optimization levels
(like -O3) which seems related to lambdas and variants.

This wrong code later leads to a crash when the browser starts at
WebCore::FrameLoader::effectiveReferrerPolicy()

* Source/WebCore/page/Page.cpp:
(WebCore::createMainFrame):

Canonical link: https://commits.webkit.org/279066@main
